datalad_next/url_operations/ssh.py

```diff
@@ -165,49 +169,56 @@ def download(self,
-            stream = props.pop('_stream')
-            expected_size = props['content-length']
-            dst_fp = sys.stdout.buffer if to_path is None \
-                else open(to_path, 'wb')
-            # Localize variable access to minimize overhead
-            dst_fp_write = dst_fp.write
-            # download can start
-            self._progress_report_start(
-                progress_id,
-                ('Download %s to %s', from_url, to_path),
-                'downloading',
-                expected_size,
-            )
-            for chunk in stream:
-                # write data
-                dst_fp_write(chunk)
-                # compute hash simultaneously
-                hasher.update(chunk)
-                self._progress_report_update(
-                    progress_id, ('Downloaded chunk',), len(chunk))
-            props.update(hasher.get_hexdigest())
-            return props
-        except CommandError as e:
-            if e.code == 244:
-                # this is the special code for a file-not-found
-                raise UrlOperationsResourceUnknown(from_url) from e
-            else:
-                # wrap this into the datalad-standard, but keep the
-                # original exception linked
-                raise UrlOperationsRemoteError(from_url, message=str(e)) from e
+            with iter_subproc(cmd) as stream:
+                props = self._get_props(from_url, stream)
+                expected_size = props['content-length']
+                # The stream might have changed due to not yet processed, but
+                # fetched data, that is now chained in front of it. Therefore we
+                # get the updated stream from the props
+                download_stream = props.pop('_stream')
+
+                dst_fp = sys.stdout.buffer \
+                    if to_path is None \
+                    else open(to_path, 'wb')
+
+                # Localize variable access to minimize overhead
+                dst_fp_write = dst_fp.write
+
+                # download can start
+                for chunk in self._with_progress(
+                        download_stream,
+                        progress_id=progress_id,
+                        label='downloading',
+                        expected_size=expected_size,
+                        start_log_msg=('Download %s to %s', from_url, to_path),
+                        end_log_msg=('Finished download',),
+                        update_log_msg=('Downloaded chunk',)
+                ):
+                    # write data
+                    dst_fp_write(chunk)
+                    # compute hash simultaneously
+                    hasher.update(chunk)
+        except IterableSubprocessError:
+            self._check_return_code(stream.returncode, from_url)
+        except StopIteration:
+            # an unexpected `StopIteration` occurred, this indicates a
+            # non-existing resource.
+            self._check_return_code(244, from_url)
```

Comment:
Why not use the actual exit code, rather than declaring one that may or may not be correct?

Reply:
If an exception is raised and the `iter_subproc`-content is left, the subprocess is terminated. A terminated `ssh` returns 255 as exit code. This code does not lead to an `UrlOperationsResourceUnknown`-exception, and would therefore not trigger the correct code-path, i.e. the "resource-not-available" code-path.

The alternative (and we can discuss whether that would be more correct) approach would be to catch `StopIteration`-exceptions (or ---if we do use `for` instead of `next()`--- detect short iterator results) and return an error indicator instead of the `props`-dictionary, e.g. `None`. That would then lead to an `IterableSubprocessError` with return code 244 (in case of non-existing files).

Reply:
OK, thanks for the explanation.